broadinstitute/single_cell_portal_core
# Automate submitting differential expression jobs on successful parse jobs (SCP-4716, SCP-4593)

#### BACKGROUND
Currently, finding candidate studies for differential expression and generating results is a completely manual process.  Admins must find potential studies that meet the criteria, then manually launch a job to generate results for eligible annotations.  This limits the reach of the differential expression feature as it requires direct intervention on our part to extend it to new studies.

#### CHANGES
This update adds automation such that on a successful clustering/metadata/expression parse job, the portal will check to see if the study is eligible for running differential expression.  The study must meet 4 criteria:
1. Public
2. Has clustering/metadata
3. Has raw count data
4. Has visualizable group-based annotations

If all 4 are true, then a job for each eligible cluster/annotation combo will be automatically submitted.  This automation will also skip any annotation/clustering combos that already have results to reduce overall load and associated costs.  The success/failure emails of these jobs also will be delivered directly to admins instead of end users, as they will not be aware that these jobs are being submitted, and could potentially be confused by those notifications.

Note: this does not automatically backfill data on studies that already exist.  These will still have to be handled in a manual fashion, though there now is the convenience method of `DifferentialExpressionService.study_eligible?` for checking if a study qualifies.

Other changes included but not part of the DE work:
* Fixed an issue when using `docker-compose` locally where emails are not delivered due to port `587` not being bound to the host, and the `DeveloperMailInterceptor` not working since `git config user.email` does not register from inside Docker.  Developers can now use the "QA Dev Email" admin configuration to ensure email delivery to their Broad email from inside Docker locally.
* Removed an extraneous tab character that caused `raw1_human_5k_cells_80_genes.raw_dense.txt` to fail validation/ingest.
* Changed `YAML.load` to `YAML.unsafe_load` in `DelayedJobAccessor` as per SCP-4593 due to persistent errors when testing this work.  Since this code cannot be called anywhere other than Rails console, there is no security issue with calling `unsafe_load`.

#### MANUAL TESTING
The simplest way to test this is to use the `SyntheticStudyPopulator` and ingest the "Human raw counts" study.  Even if you already have this study locally, the populator will remove and re-ingest all files, thus triggering the automation. 
1. Boot as normal, and start DelayedJob (using `bin/docker-compose-setup.sh` will automate all of this if you aren't already doing so)
2. In a separate terminal, open a Rails console session and run the following (you do not need to attach to the Docker instance to do so, though that will also work):
```
SyntheticStudyPopulator.populate('human_raw_counts')
```
3. In `development.log`, you should see numerous instances of PAPI jobs being launched to ingest files
4. In 2-3 minutes, as files begin to finish parsing, you should start to see messages denoting that differential expression jobs are being launched.  The exact number of jobs will vary on the order in which files finish parsing, but there can be a maximum of 4.  In the example below, the 2D cluster file failed to ingest due to a transient MongoDB error, so that cluster was skipped.  This demonstrates both the automation and error handling for differential expression jobs:
```
SCP79 is eligible for differential expression, launching available jobs
SCP79 has annotations eligible for DE; validating inputs
Checking DE job for SCP79: raw1_human_5k_cells_80_genes.cluster_2D.txt (CLUSTER--group--study)
  Skipping DE job for SCP79: raw1_human_5k_cells_80_genes.cluster_2D.txt (CLUSTER--group--study) due to: cannot find cluster for raw1_human_5k_cells_80_genes.cluster_2D.txt
Checking DE job for SCP79: raw1_human_5k_cells_80_genes.cluster_2D.txt (SUBCLUSTER--group--study)
  Skipping DE job for SCP79: raw1_human_5k_cells_80_genes.cluster_2D.txt (SUBCLUSTER--group--study) due to: cannot find cluster for raw1_human_5k_cells_80_genes.cluster_2D.txt
Checking DE job for SCP79: raw1_human_5k_cells_80_genes.cluster_3D.txt (CLUSTER--group--study)
==> DE job for SCP79: raw1_human_5k_cells_80_genes.cluster_3D.txt (CLUSTER--group--study) successfully launched
Checking DE job for SCP79: raw1_human_5k_cells_80_genes.cluster_3D.txt (SUBCLUSTER--group--study)
Remote found for raw1_human_5k_cells_80_genes.cluster_3D.txt:6362d0ac94ec8f12627bf512, launching Ingest job
==> DE job for SCP79: raw1_human_5k_cells_80_genes.cluster_3D.txt (SUBCLUSTER--group--study) successfully launched
SCP79 yielded 2 differential expression jobs; 2 skipped
```
5.  Once the DE jobs finish, you should be able to navigate to the Explore tab and view them.  This is a screenshot of the 3D cluster viewing the `SUBCLUSTER` annotation:
![Screen Shot 2022-11-02 at 4 31 18 PM](https://user-images.githubusercontent.com/729968/199597541-255085cd-b922-4d10-8f70-e8e9dd9bef2a.png)
